pkg/node/node_connectivity_checker.go

```diff
@@ -0,0 +1,275 @@
+	if !s.isPodmonEnabled {
+		log.Info("podmon is not enabled")
+		return
+	}
+	pollingFrequencyInSeconds = common.SetPollingFrequency(ctx)
+	setAPIPort(ctx)
+	s.startNodeToArrayConnectivityCheck(ctx)
+	s.apiRouter(ctx)
+}
+
+// setAPIPort set the port for running server
+func setAPIPort(ctx context.Context) {
+	if port, ok := csictx.LookupEnv(ctx, common.EnvPodmonAPIPORT); ok && strings.TrimSpace(port) != "" {
+		apiPort = fmt.Sprintf(":%s", port)
+		log.Debugf("set podmon API port to %s", apiPort)
+		return
+	}
+	// If the port number cannot be fetched, set it to default
+	apiPort = ":" + common.DefaultPodmonAPIPortNumber
+	log.Debugf("set podmon API port to default %s", apiPort)
+}
+
+// apiRouter serves http requests
+func (s *Service) apiRouter(ctx context.Context) {
+	log.Infof("starting http server on port %s", apiPort)
+	// create a new mux router
+	router := mux.NewRouter()
+	// route to connectivity status
+	// nodeHealth & connectivityStatus are the handlers
+	router.HandleFunc(nodeStatus, nodeHealth).Methods("GET")
+	router.HandleFunc(arrayStatus, connectivityStatus).Methods("GET")
+	router.HandleFunc(arrayStatus+"/"+"{arrayId}", getArrayConnectivityStatus).Methods("GET")
+	// start http server to serve requests
+	server := &http.Server{
+		Addr:         apiPort,
+		Handler:      router,
+		ReadTimeout:  common.Timeout,
+		WriteTimeout: common.Timeout,
+	}
+	err := server.ListenAndServe()
+	if err != nil {
+		log.Errorf("unable to start http server to serve status requests due to %s", err)
+	}
+}
+
+// nodeHealth handler states if the node is up
+func nodeHealth(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+	w.Header().Set("Content-Type", "application/json")
+	fmt.Fprintf(w, "node is up and running \n")
+}
+
+// connectivityStatus handler returns array connectivity status
+func connectivityStatus(w http.ResponseWriter, r *http.Request) {
+	log.Infof("connectivityStatus called, status is %v \n", probeStatus)
+	// w.Header().Set("Content-Type", "application/json")
+	if probeStatus == nil {
+		log.Errorf("error probeStatus map in cache is empty")
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Header().Set("Content-Type", "application/json")
+		return
+	}
+
+	// convert struct to JSON
+	log.Debugf("ProbeStatus fetched from the cache has %+v", probeStatus)
+
+	jsonResponse, err := MarshalSyncMapToJSON(probeStatus)
+	if err != nil {
+		log.Errorf("error %s during marshaling to json", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Header().Set("Content-Type", "application/json")
+		return
+	}
+	log.Info("sending connectivityStatus for all arrays ")
+	_, err = w.Write(jsonResponse)
+	if err != nil {
+		w.WriteHeader(http.StatusOK)
+		w.Header().Set("Content-Type", "application/json")
+		log.Errorf("unable to write response %s", err)
+	}
+}
+
+// MarshalSyncMapToJSON marshal the sync Map to Json
+func MarshalSyncMapToJSON(m *sync.Map) ([]byte, error) {
```

Comment:
Suggestion for future improvements: If these functions are duplicated in each driver code to run a HTTP server and handle HTTP requests, we can move these common functions to a seperate library.

Reply:
Sure Priya. Nice idea.